File: 725.split-linked-list-in-parts.ts
/*
 * @lc app=leetcode id=725 lang=typescript
 *
 * [725] Split Linked List in Parts
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

class ListNode725 {
  val: number;
  next: ListNode725 | null;
  constructor(val?: number, next?: ListNode725 | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}
function splitListToParts(
  head: ListNode725 | null,
  k: number
): Array<ListNode725 | null> {
  let len = 0;
  let current = head;

  while (current !== null) {
    len++;
    current = current.next;
  }

  const list: Array<ListNode725 | null> = [];

  for (let i = 0; i < k; i++) {
    list.push(null);
  }
  current = head;
  let prev = head;

  const fullS = Math.floor(len / k);
  let remS = len % k;
  let i = 0;

  while (current !== null) {
    list[i] = current;

    if (remS > 0) {
      prev = current;
      current = current.next;
      remS--;
    }

    for (let j = 0; j < fullS; j++) {
      prev = current;
      current = current.next;
    }

    prev.next = null;
    prev = current;
    i++;
  }

  return list;
}
// @lc code=end

File: 725.split-linked-list-in-parts.rb
#
# @lc app=leetcode id=725 lang=ruby
#
# [725] Split Linked List in Parts
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @param {Integer} k
# @return {ListNode[]}
def split_list_to_parts(head, k)
  list = Array.new(k) { nil }

  current = head
  len = 0

  until current.nil?
    current = current.next
    len += 1
  end

  whole = len / k
  rem = len % k

  current = head
  prev = head
  i = 0

  until current.nil?
    list[i] = current

    (0...whole).each do
      prev = current
      current = current.next
    end

    if rem > 0
      rem -= 1
      prev = current
      current = current.next
    end

    prev.next = nil
    prev = current
    i += 1
  end

  list
end
# @lc code=end
